doc/SAI-Proposal-VlanStack-QinQ.md

```diff
@@ -0,0 +1,380 @@
+
+ Title       | Vlan Stack
+-------------|-----------------------------------------------------------------
+ Authors     | Edge-Core
+ Status      | In review
+ Type        | Standards track
+ Created     | 25/02/2022
+ Updated     | -
+ SAI-Version | 1.9.1
+
+-------------------------------------------------------------------------------
+
+## Overview ##
+
+The Vlan Stack feature includes two functions, vlan translation and QinQ.
+
+![qinq](figures/vlan_stack/vlan_stack_QinQ_usecase.png "Figure 1: QinQ usage scenario")
+__Figure 1: QinQ usage scenario__
+
+![vlan_translate](figures/vlan_stack/vlan_stack_vlan_xlate_usecase.png "Figure 2: Vlan translation usage scenario")
+__Figure 2: Vlan Translation usage scenario__
+
+## Packet Format ##
+
+Untagged packet, Single tagged packet and Double tagged packet
+
+![packet_format](figures/vlan_stack/TCPIP_802.1ad_DoubleTag.png "Figure 3: Packet packet format")
+__Figure 3: Packet format__
+
+## SAI Pipeline
+
+The packet processing pipeline is below:
+![vlan_stack-pipeline](figures/vlan_stack/vlan_stack_pipeline.png "Figure 4: Vlan Stack SAI Pipeline")
+__Figure 4: Vlan Stack SAI Pipeline__
+
+## SAI Attribute
+
+| SAI Component | SAI attribute                                  | Description                     |
+| ------------- | ---------------------------------------------- | ------------------------------- |
+| VLAN Stacking | SAI_VLAN_STACK_ATTR_STAGE                      | At which packet flow direction the action should take effect. Ingress or Egress    |
+|               | SAI_VLAN_STACK_ATTR_ACTION                     | Include SWAP, PUSH and POP operation    |
+|               | SAI_VLAN_STACK_ATTR_MATCH_TYPE                 | At which dot1q header position should be used to match original vlan ID. Inner or Outer. Please note that for a PUSH action, this should always set to Inner.    |
+|               | SAI_VLAN_STACK_ATTR_VLAN_APPLIED_PRI           | Rule priority from low to high can assign this from 0 to 7    |
+|               | SAI_VLAN_STACK_ATTR_ORIGINAL_VLAN_ID_LIST      | Action will take effect if vlan ID of a packet match any one of this list    |
+|               | SAI_VLAN_STACK_ATTR_PORT                       | Target port ID                  |
+|               | SAI_VLAN_STACK_ATTR_APPLIED_VLAN_ID            | New vlan ID                     |
+
+
+## Use cases
+
+### Ingress/Add outer tag based on inner tag
+Add an outer dot1q header with VID=100 for those packets with VID=10 at Ethernet10
+
+#### Usage
+
+```
+sai_object_id_t vlan_stacking_oid;
+sai_attribute_t attr;
+vector<sai_attribute_t> vlan_stacking_entry_attrs;
+
+attr.id = SAI_VLAN_STACK_ATTR_STAGE;
+attr.value.s32 = SAI_VLAN_STACK_STAGE_INGRESS;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_ACTION;
+attr.value.s32 = SAI_VLAN_STACK_ACTION_PUSH;
+vlan_stacking_entry_attrs.push(attr);
+
+attr.id = SAI_VLAN_STACK_ATTR_MATCH_TYPE;
+attr.value.s32 = SAI_VLAN_STACK_MATCH_TYPE_INNER;
```

Comment:
Should the match type be outer for the ingress stage? For example, in the 2nd row of the table below, we are matching on the outer tag (svid=10) of the packet.

Reply:
I know this point will confuse us easily. So, I added some notes at #560 of inc/saivlan.h
```
    /**
     * @brief Match type
     * The original Vlan ID should match to inner tag or outer tag
     * For PUSH action, this should always set to INNER. Since compare to the new
     * outer tag, the original one is an inner tag.
     *
     * @type sai_vlan_stack_match_type_t
     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
     */
    SAI_VLAN_STACK_ATTR_MATCH_TYPE,
```

This design also make us simple to implement. But I think this can open for discussion.

Reply:
So, you are saying that match_type is based on how the packet looks like **after** all the stacking operations have been performed? Do we have a case where match_type will be outer?

Reply:
Not really at all.
This is what I think.
The PUSH action usually be used in QinQ case. When a packet come in, a new vlan tag will be added based on the inner tag. And when a packet go out, the vlan tag will be removed based on the outer tag. Doesn't this sounds more reasonable?

And for SWAP action. I can swap the inner vlan tag or the outer vlan tag.

Reply:
This might be applicable, when we match a packet with double-tag. Otherwise, we should always be matching the outer-tag?

Reply:
You two persuade me. I will change this case to match outer in the next commit.